pandas/tests/internals/test_internals.py

```diff
@@ -1242,3 +1242,34 @@ def test_interleave_non_unique_cols():
     assert df_unique.values.shape == df.values.shape
     tm.assert_numpy_array_equal(df_unique.values[0], df.values[0])
     tm.assert_numpy_array_equal(df_unique.values[1], df.values[1])
+
+
+def test_assign_df_datetime_column_with_index():
+    df = pd.DataFrame(
+        [
+            [datetime(2020, 2, 28, 13, 51, 27, tzinfo=timezone.utc)],
+            [datetime(2020, 2, 28, 13, 51, 27, tzinfo=timezone.utc)],
+            [datetime(2020, 2, 28, 13, 51, 27, tzinfo=timezone.utc)],
+            [datetime(2020, 2, 28, 13, 51, 27, tzinfo=timezone.utc)],
+        ],
+        columns=["data"],
+    )
+    df2 = pd.DataFrame(index=df.index)
+    df2.loc[df.index, "data"] = df["data"]
+    tm.assert_numpy_array_equal(np.array(df.data), np.array(df2.data))
```

Comment:
`tm.assert_frame_equal(df, df2)`

Reply:
`df.data` has type `datetime[ns, tz]`, `df2.data` has type `object`

Reply:
Isn't that a bug in and of itself? I think dtypes should also match

Reply:
That is consistent with the behaviour in 0.25. If you have are using an index assignment the fill values are NaT, so the type after assignment is `object`. I am not sure if there is a better way to handle this.